File: solutions/go/yacht/1/yacht.go
package yacht

func sumOccurrences(dice []int, eyes int) (result int) {
	for _, element := range dice {
		if element == eyes {
			result += eyes
		}
	}
	return
}

func countFourOfAKind(dice []int) int {
	for i := 1; i <= 6; i++ {
		count := 0
		for _, die := range dice {
			if die == i {
				count++
			}
		}
		if count >= 4 {
			return 4 * i
		}
	}
	return 0
}

func countYacht(dice []int) int {
	for i := 1; i <= 6; i++ {
		count := 0
		for _, die := range dice {
			if die == i {
				count++
			}
		}
		if count == 5 {
			return 50
		}
	}
	return 0
}

func countFullHouse(dice []int) int {
	hasTwoDice := false
	hasThreeDice := false
	for i := 1; i <= 6; i++ {
		count := 0
		for _, die := range dice {
			if die == i {
				count++
			}
		}
		if count == 2 {
			hasTwoDice = true
			continue
		}
		if count == 3 {
			hasThreeDice = true
			continue
		}
	}
	if hasTwoDice && hasThreeDice {
		return sum(dice, len(dice))
	}
	return 0
}

func countLittleStraight(dice []int) int {
	for i := 1; i < 6; i++ {
		found := false
		for _, element := range dice {
			if element == i {
				found = true
				break
			}
		}
		if !found {
			return 0
		}
	}
	return 30
}

func countBigStraight(dice []int) int {
	for i := 2; i <= 6; i++ {
		found := false
		for _, element := range dice {
			if element == i {
				found = true
				break
			}
		}
		if !found {
			return 0
		}
	}
	return 30
}

func sum(slice []int, n int) int {
	if n <= 0 {
		return 0
	}
	return sum(slice, n-1) + slice[n-1]
}

func Score(dice []int, category string) int {
	switch category {
	case "ones":
		return sumOccurrences(dice, 1)
	case "twos":
		return sumOccurrences(dice, 2)
	case "threes":
		return sumOccurrences(dice, 3)
	case "fours":
		return sumOccurrences(dice, 4)
	case "fives":
		return sumOccurrences(dice, 5)
	case "sixes":
		return sumOccurrences(dice, 6)
	case "choice":
		return sum(dice, len(dice))
	case "four of a kind":
		return countFourOfAKind(dice)
	case "full house":
		return countFullHouse(dice)
	case "little straight":
		return countLittleStraight(dice)
	case "big straight":
		return countBigStraight(dice)
	case "yacht":
		return countYacht(dice)
	}
	return 0
}

File: solutions/javascript/yacht/1/yacht.js
/**
 * Calculates the score for the Yacht category.
 * @param {number[]} dices - Array of dice values.
 * @returns {number} - The score for the Yacht category.
 */
const getScoreForTheYachtCategory = (dices) => {
  const isYacht = new Set(dices).size === 1;
  return isYacht ? 50 : 0;
};

/**
 * Maps dice values to a counter array.
 * @param {number[]} dices - Array of dice values.
 * @returns {number[]} - Counter array for dice values.
 */
const mapDicesToCounterArray = (dices) => {
  let counterArray = [0, 0, 0, 0, 0, 0];
  for (let item of dices) {
    counterArray[item - 1]++;
  }
  return counterArray;
};

/**
 * Gets the number of appearances of a specific dice value.
 * @param {number[]} dices - Array of dice values.
 * @param {number} diceValue - The dice value to count.
 * @returns {number} - The number of appearances of the dice value.
 */
const getNoOfAppearances = (dices, diceValue) => {
  return dices.filter((value) => value === diceValue).length;
};

/**
 * Calculates the sum of dice values.
 * @param {number[]} dices - Array of dice values.
 * @returns {number} - The sum of dice values.
 */
const getSumOfDices = (dices) => {
  return dices.reduce((a, b) => a + b, 0);
};

/**
 * Calculates the score for the Four of a Kind category.
 * @param {number[]} dices - Array of dice values.
 * @returns {number} - The score for the Four of a Kind category.
 */
const getScoreForTheFourOfAKindCategory = (dices) => {
  const counterArray = mapDicesToCounterArray(dices);
  for (let i = 0; i < counterArray.length; i++) {
    if (counterArray[i] >= 4) {
      return 4 * (i + 1);
    }
  }
  return 0;
};

/**
 * Calculates the score for the Little Straight category.
 * @param {number[]} dices - Array of dice values.
 * @returns {number} - The score for the Little Straight category.
 */
const getScoreForTheLittleStraightCategory = (dices) => {
  const counterArray = mapDicesToCounterArray(dices);
  const isLittleStraight = arrayIsFilledWithValue(
    counterArray,
    0,
    counterArray.length - 1,
    1,
  );
  return isLittleStraight ? 30 : 0;
};

/**
 * Calculates the score for the Big Straight category.
 * @param {number[]} dices - Array of dice values.
 * @returns {number} - The score for the Big Straight category.
 */
const getScoreForTheBigStraightCategory = (dices) => {
  const counterArray = mapDicesToCounterArray(dices);
  const isBigStraight = arrayIsFilledWithValue(
    counterArray,
    1,
    counterArray.length,
    1,
  );
  return isBigStraight ? 30 : 0;
};

/**
 * Checks if an array is filled with a specific value within a range.
 * @param {number[]} array - Array to check.
 * @param {number} startPos - Start position in the array.
 * @param {number} endPos - End position in the array.
 * @param {number} value - Value to check for.
 * @returns {boolean} - True if the array is filled with the value, false otherwise.
 */
const arrayIsFilledWithValue = (array, startPos, endPos, value) => {
  for (let i = startPos; i < endPos; i++) {
    if (array[i] !== value) {
      return false;
    }
  }
  return true;
};

/**
 * Calculates the score for the Full House category.
 * @param {number[]} dices - Array of dice values.
 * @returns {number} - The score for the Full House category.
 */
const getScoreForTheFullHouseCategory = (dices) => {
  const counterArray = mapDicesToCounterArray(dices);
  let hasTwoOfAKind = false;
  let hasThreeOfAKind = false;
  for (let item of counterArray) {
    if (item === 2) {
      hasTwoOfAKind = true;
    } else if (item === 3) {
      hasThreeOfAKind = true;
    }
  }
  return hasTwoOfAKind && hasThreeOfAKind ? getSumOfDices(dices) : 0;
};

/**
 * Calculates the score based on the dice values and the category.
 * @param {number[]} dices - Array of dice values.
 * @param {string} category - Category to calculate the score for.
 * @returns {number} - The calculated score.
 */
export const score = (dices, category) => {
  switch (category) {
    case 'yacht':
      return getScoreForTheYachtCategory(dices);
    case 'ones':
      return getNoOfAppearances(dices, 1);
    case 'twos':
      return 2 * getNoOfAppearances(dices, 2);
    case 'threes':
      return 3 * getNoOfAppearances(dices, 3);
    case 'fours':
      return 4 * getNoOfAppearances(dices, 4);
    case 'fives':
      return 5 * getNoOfAppearances(dices, 5);
    case 'sixes':
      return 6 * getNoOfAppearances(dices, 6);
    case 'full house':
      return getScoreForTheFullHouseCategory(dices);
    case 'four of a kind':
      return getScoreForTheFourOfAKindCategory(dices);
    case 'little straight':
      return getScoreForTheLittleStraightCategory(dices);
    case 'big straight':
      return getScoreForTheBigStraightCategory(dices);
    case 'choice':
      return getSumOfDices(dices);
  }
  return 0;
};
